crypto/rsa/rsa_pss.c

```diff
@@ -168,19 +169,28 @@ int RSA_padding_add_PKCS1_PSS_mgf1(RSA *rsa, unsigned char *EM,
         goto err;
     }
 
-    MSBits = (BN_num_bits(rsa->n) - 1) & 0x7;
     emLen = RSA_size(rsa);
+    if (emLen <= 1)
+        goto err;
+
+    MSBits = (BN_num_bits(rsa->n) - 1) & 0x7;
     if (MSBits == 0) {
         *EM++ = 0;
         emLen--;
     }
     if (sLen == RSA_PSS_SALTLEN_MAX) {
+        if (hLen < 2 || hLen - 2 > emLen)
+            goto err;
         sLen = emLen - hLen - 2;
-    } else if (emLen < (hLen + sLen + 2)) {
+    } else if (sLen > INT_MAX - 2 - hLen ||
+               emLen < (hLen + sLen + 2)) {
         RSAerr(RSA_F_RSA_PADDING_ADD_PKCS1_PSS_MGF1,
-               RSA_R_DATA_TOO_LARGE_FOR_KEY_SIZE);
+                RSA_R_DATA_TOO_LARGE_FOR_KEY_SIZE);
```

Comment:
Doesn't now look the RSA_padding_add_PKCS1_PSS_mgf1 look completely different than
RSA_verify_PKCS1_PSS_mgf1 ?